### ISSUE TITLE
Metal resource leak

### ISSUE THREAD

**Issue by dan-j-cooper** [waybarrios/vllm-mlx#91]
When running lm_eval to get some points of comparison between different quantized models I ran into a resource leak which ground the gui of my mbp to a standstill as all available metal resources were exhausted. 

It looks like for whatever reason the system is accumulating metal buffers and never evaluating them, so eventually the system gets depleted. 
I think it’s plausible that the logits only being conditionally evaluated in _step is the culprit, but I didn’t want to go digging down into that too much.

I wrote a script(below) to run some concurrent loading at the server and you can observe the number of handles monotonically increasing. 
A fix that’s relatively simple and low overhead is adding the following to the scheduler.

```
2231 2231:         # Clear finished tracking for next step
2232 2232:         old_finished = self.finished_req_ids
2233 2233:         self.finished_req_ids = set()
     2234:         # Adaptive interval: scale inversely with concurrency to prevent
     2235:         # Metal resource handle exhaustion under high-concurrency workloads.
     2236:         active_seqs = len(self.running)
     2237:         effective_interval = max(
     2238:             8, self._clear_cache_interval // max(1, active_seqs // 8)
     2239:         )
2234 2240:
2235 2241:         # Periodically clear Metal cache to prevent memory accumulation
2236 2242:         self._step_count += 1
2237 2243:         if self._step_count % effective_interval == 0:
     2244:             # Evaluate batch.tokens to collapse lazy concatenation chains
     2245:             if (
     2246:                 self.batch_generator is not None
     2247:                 and self.batch_generator.active_batch is not None
     2248:             ):
     2249:                 tokens = self.batch_generator.active_batch.tokens
     2250:                 if tokens:
     2251:                     mx.eval(*tokens)
2238 2252:             mx.clear_cache()
```

Note that the default model I was using was qwen3-coder-next, but I observed the same resource leak using glm 4.5.
```
from __future__ import annotations

import argparse
import asyncio
import contextlib
import dataclasses
import math
import re
import signal
import subprocess
import sys
import time
from pathlib import Path

import numpy as np
import requests

# ── prompts (varied to avoid prefix-cache hits) ─────────────────────────────

PROMPTS = [
    "Explain how a hash table works in simple terms.",
    "Write a Python function that checks if a string is a palindrome.",
    "What are the main differences between TCP and UDP?",
    "Describe the water cycle in three sentences.",
    "Write a haiku about programming.",
    "Explain the concept of recursion with an example.",
    "What is the difference between a stack and a queue?",
    "Summarize the plot of Romeo and Juliet in two sentences.",
    "Write a bash one-liner to find the largest files in a directory.",
    "Explain what an API is to a non-technical person.",
]

# ── data layout + presets ────────────────────────────────────────────────────

COL = {"t": 0, "agx": 1, "iobufs": 2, "tok_s": 3, "running": 4, "waiting": 5, "done": 6}

_IOCLS_RE = re.compile(r"(\w+)\s*=\s*(\d+)")


@dataclasses.dataclass(frozen=True, slots=True)
class Preset:
    concurrency: int
    duration: int
    max_tokens: int
    warmup: float
    interval: float
    server_args: str


SHORT = Preset(
    concurrency=16, duration=120, max_tokens=256,
    warmup=30.0, interval=5.0,
    server_args="--continuous-batching",
)
LONG = Preset(
    concurrency=32, duration=300, max_tokens=2048,
    warmup=60.0, interval=3.0,
    server_args=(
        "--continuous-batching --enable-prefix-cache --max-num-seqs 64"
        " --prefill-batch-size 64 --completion-batch-size 64"
        " --prefix-cache-size 1024 --use-paged-cache"
        " --paged-cache-block-size 64 --max-cache-blocks 8192"
        " --stream-interval 8 --max-tokens 128000"
    ),
)

# ── ioclasscount (async subprocess) ─────────────────────────────────────────


async def _ioclasscount() -> tuple[int, int]:
    """Return (AGXAllocation, IOGPUBufferMemoryDescriptor). Raises on failure."""
    proc = await asyncio.create_subprocess_exec(
        "ioclasscount", "AGXAllocation", "IOGPUBufferMemoryDescriptor",
        stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE,
    )
    stdout, _ = await asyncio.wait_for(proc.communicate(), timeout=5)
    if proc.returncode != 0:
        raise RuntimeError(f"ioclasscount exited {proc.returncode}")
    vals = dict((k, int(v)) for k, v in _IOCLS_RE.findall(stdout.decode()))
    return vals["AGXAllocation"], vals["IOGPUBufferMemoryDescriptor"]


# ── server context manager ───────────────────────────────────────────────────


@contextlib.contextmanager
def managed_server(model: str, port: int, server_args: str):
    """Launch vllm-mlx, wait for /health, yield proc, cleanup on exit."""
    cmd = [
        sys.executable, "-m", "vllm_mlx.cli", "serve", model,
        "--host", "0.0.0.0", "--port", str(port),
        *server_args.split(),
    ]
    log_path = Path("measurements") / "server.log"
    log_path.parent.mkdir(parents=True, exist_ok=True)
    log_file = log_path.open("w")

    print(f"Starting: {' '.join(cmd)}")
    proc = subprocess.Popen(cmd, stdout=log_file, stderr=subprocess.STDOUT)
    try:
        base = f"http://localhost:{port}"
        deadline = time.monotonic() + 120
        print("Waiting for server...", end="", flush=True)
        while time.monotonic() < deadline:
            try:
                if requests.get(f"{base}/health", timeout=3).status_code == 200:
                    print(" ready!")
                    yield proc
                    return
            except requests.ConnectionError:
                pass
            if proc.poll() is not None:
                sys.exit(f"\nServer exited early (code {proc.returncode})")
            print(".", end="", flush=True)
            time.sleep(2)
        sys.exit("\nServer failed to start within 120s")
    finally:
        proc.terminate()
        try:
            proc.wait(timeout=10)
        except subprocess.TimeoutExpired:
            proc.kill()
        log_file.close()


# ── worker ───────────────────────────────────────────────────────────────────


async def _worker(
    session: requests.Session, url: str, model: str,
    max_tokens: int, counters: list[int], stop: asyncio.Event,
) -> None:
    """Fire requests in a loop. counters = [done, errs, toks]."""
    idx = 0
    while not stop.is_set():
        payload = {
            "model": model,
            "messages": [{"role": "user", "content": PROMPTS[idx % len(PROMPTS)]}],
            "max_tokens": max_tokens, "temperature": 0.7, "stream": False,
        }
        try:
            r = await asyncio.to_thread(
                session.post, f"{url}/v1/chat/completions", json=payload, timeout=120,
            )
            if r.status_code == 200:
                counters[0] += 1
                counters[2] += r.json().get("usage", {}).get("completion_tokens", 0)
            else:
                counters[1] += 1
        except Exception:
            counters[1] += 1
            try:
                await asyncio.wait_for(stop.wait(), timeout=2)
            except asyncio.TimeoutError:
                pass
        idx += 1


# ── main loop ────────────────────────────────────────────────────────────────

HEADER = f"{'time':>6}  {'AGX':>9}  {'IOBufs':>9}  {'tok/s':>7}  {'run':>3} {'wait':>4}  {'done':>5}"


async def run(
    model: str, port: int, preset: Preset, proc: subprocess.Popen | None,
) -> None:
    url = f"http://localhost:{port}"
    log_path = Path("measurements") / "server.log"

    cap = math.ceil(preset.duration / preset.interval) + 2
    data = np.full((cap, len(COL)), np.nan)
    row = 0
    prev_toks: int | None = None

    counters = [0, 0, 0]  # done, errs, toks
    stop = asyncio.Event()

    session = requests.Session()
    tasks = [
        asyncio.create_task(
            _worker(session, url, model, preset.max_tokens, counters, stop)
        )
        for _ in range(preset.concurrency)
    ]

    def _sigint(*_: object) -> None:
        print("\nStopping...")
        stop.set()
    signal.signal(signal.SIGINT, _sigint)

    t0 = time.monotonic()
    print(f"\n{HEADER}\n{'-' * 72}")

    try:
        while not stop.is_set():
            elapsed = time.monotonic() - t0
            if elapsed >= preset.duration:
                break

            # sample ioclasscount — exit loop on failure
            try:
                agx, iobufs = await _ioclasscount()
            except Exception as e:
                print(f"\nioclasscount failed: {e}")
                break

            # sample /v1/status — exit loop if server unreachable
            try:
                r = await asyncio.to_thread(
                    session.get, f"{url}/v1/status", timeout=5,
                )
                r.raise_for_status()
                d = r.json()
                running = d.get("num_running", 0)
                waiting = d.get("num_waiting", 0)
                total_toks = d.get("total_completion_tokens", 0)
            except Exception as e:
                print(f"\nServer heartbeat failed: {e}")
                break

            # tok/s since last sample
            tok_s_val = np.nan
            if prev_toks is not None and row > 0:
                dt = elapsed - data[row - 1, COL["t"]]
                if dt > 0:
                    tok_s_val = (total_toks - prev_toks) / dt
            prev_toks = total_toks

            # fill row
            if row < cap:
                data[row] = [elapsed, agx, iobufs, tok_s_val, running, waiting, counters[0]]

            # print line
            ts_s = f"{tok_s_val:>7.0f}" if not np.isnan(tok_s_val) else f"{'---':>7}"
            print(
                f"{elapsed:>5.0f}s  {agx:>9,}  {iobufs:>9,}  {ts_s}  "
                f"{running:>3} {waiting:>4}  {counters[0]:>5}",
                flush=True,
            )

            # check server alive
            if proc and proc.poll() is not None:
                tail = "\n".join(log_path.read_text().splitlines()[-20:]) if log_path.exists() else ""
                print(f"\n  SERVER DIED (exit {proc.returncode}) at {elapsed:.0f}s")
                if tail:
                    print(f"  Last 20 lines of server.log:\n{tail}")
                break

            row += 1
            nxt = t0 + row * preset.interval
            try:
                await asyncio.wait_for(stop.wait(), timeout=max(0, nxt - time.monotonic()))
            except asyncio.TimeoutError:
                pass
    finally:
        stop.set()
        for t in tasks:
            t.cancel()
        await asyncio.gather(*tasks, return_exceptions=True)
        session.close()

    # ── summary ──────────────────────────────────────────────────────────
    d = data[:row]
    if row == 0:
        print("No samples collected.")
        return

    actual = time.monotonic() - t0
    lines = [
        f"\n{'=' * 60}",
        "  Metal Handle Leak Measurement",
        f"{'=' * 60}",
        f"  Model:       {model}",
        f"  Duration:    {actual:.0f}s   Concurrency: {preset.concurrency}",
        f"  Requests:    {counters[0]} ok, {counters[1]} errors, {counters[2]} tokens",
    ]

    for label, ci in [("AGXAllocation", COL["agx"]), ("IOGPUBufMemDesc", COL["iobufs"])]:
        col = d[:, ci]
        valid = ~np.isnan(col)
        if not valid.any():
            continue
        first, last = col[valid][0], col[valid][-1]
        delta = last - first
        lines.append(f"\n  {label}:  {first:,.0f} -> {last:,.0f}   delta: {delta:+,.0f}")

        mask = valid & (d[:, COL["t"]] >= preset.warmup)
        post = d[mask]
        if len(post) >= 3:
            slope = float(np.polyfit(post[:, COL["t"]] / 60, post[:, ci], 1)[0])
            lines.append(f"    slope: {slope:+,.0f} /min (post-warmup)")
            if ci == COL["agx"]:
                if slope > 100:
                    lines.append(f"    >>> LEAK: {slope:+,.0f} handles/min")
                elif slope <= 10:
                    lines.append(f"    >>> STABLE ({slope:+,.0f}/min)")

    print("\n".join(lines) + "\n")


# ── CLI ──────────────────────────────────────────────────────────────────────


def main() -> None:
    p = argparse.ArgumentParser(description="Measure Metal handle leak under load")
    p.add_argument("model", nargs="?", default="mlx-community/Qwen3-0.6B-4bit")
    p.add_argument("--port", type=int, default=8000)
    p.add_argument("--long", action="store_true", help="5 min, c=32, paged-cache stress")
    p.add_argument("--no-start-server", action="store_true", help="use existing server")
    args = p.parse_args()

    preset = LONG if args.long else SHORT

    if args.no_start_server:
        url = f"http://localhost:{args.port}"
        print(f"Connecting to existing server at {url}...")
        try:
            r = requests.get(f"{url}/health", timeout=10)
            r.raise_for_status()
        except Exception as e:
            sys.exit(f"Cannot reach server: {e}")
        asyncio.run(run(args.model, args.port, preset, proc=None))
    else:
        with managed_server(args.model, args.port, preset.server_args) as proc:
            asyncio.run(run(args.model, args.port, preset, proc=proc))


if __name__ == "__main__":
    main()
```
